lib/pact/api/exec_command_request.ex

```diff
@@ -0,0 +1,315 @@
+
+  defp validate_keypair(%KeyPair{
+         pub_key: pub_key_arg,
+         secret_key: secret_key_arg,
+         clist: clist_arg
+       }) do
+    with {:ok, pub_key} <- validate_key({:pub_key, pub_key_arg}),
+         {:ok, secret_key} <- validate_key({:secret_key, secret_key_arg}),
+         {:ok, clist} <- validate_optional_caps_list({:clist, clist_arg}) do
+      %KeyPair{pub_key: pub_key, secret_key: secret_key, clist: clist}
+    end
+  end
+
+  @spec validate_key(arg :: arg()) :: arg_validation()
+  defp validate_key({_arg, key}) when is_binary(key) and byte_size(key) == 64, do: {:ok, key}
+  defp validate_key({arg, _key}), do: {:error, [{arg, :invalid}]}
+
+  @spec validate_optional_caps_list(arg :: arg()) :: arg_validation()
+  defp validate_optional_caps_list({arg, clist}) do
+    case clist do
+      %OptionalCapsList{} = optional_caps -> {:ok, optional_caps}
+      %CapsList{} = caps_list -> {:ok, OptionalCapsList.new(caps_list)}
+      _error -> {:error, [{arg, :invalid}]}
+    end
+  end
+
+  @spec create_payload(code :: code(), data :: data()) :: valid_payload()
+  defp create_payload(code, data) do
+    [code: code, data: data]
+    |> ExecPayload.new()
+    |> PactPayload.new()
+    |> (&{:ok, &1}).()
+  end
+
+  @spec create_command(
+          hash :: hash(),
+          sigs :: signatures(),
+          cmd :: cmd()
+        ) :: valid_command()
+  defp create_command(hash, sigs, cmd) do
+    case Command.new(hash: hash, sigs: sigs, cmd: cmd) do
+      %Command{} = command -> {:ok, command}
+    end
+  end
+
+  @spec command_to_json_string(payload :: pact_payload(), t()) :: valid_command_json_string()
+  defp command_to_json_string(payload, %__MODULE__{
+         network_id: network_id,
+         meta_data: meta_data,
+         signers: signers,
+         nonce: nonce
+       }) do
+    [
+      network_id: network_id,
+      payload: payload,
+      meta: meta_data,
+      signers: signers,
+      nonce: nonce
+    ]
+    |> CommandPayload.new()
+    |> JSONPayload.parse()
+    |> (&{:ok, &1}).()
+  end
+
+  @spec sign_commands(signs :: list(), cmd :: json_string_payload(), keypairs()) ::
+          valid_sign_commands()
+  defp sign_commands(signs, _cmd, []), do: {:ok, signs}
+
+  defp sign_commands(signs, cmd, [%KeyPair{} = keypair | keypairs]) do
+    signs
+    |> sign_command(cmd, keypair)
+    |> sign_commands(cmd, keypairs)
+  end
+
+  @spec sign_command(signs :: list(), cmd :: json_string_payload(), keypair()) ::
+          list()
+  defp sign_command(signs, cmd, %KeyPair{} = keypair) do
+    {:ok, sign_command} = Sign.sign(cmd, keypair)
+    signs ++ [sign_command]
+  end
+
+  @spec get_signatures(sign_commands :: sign_commands(), result :: list()) :: valid_signatures()
+  defp get_signatures([], result), do: {:ok, SignaturesList.new(result)}
+
+  defp get_signatures([%SignCommand{sig: sig} | rest], result),
+    do: get_signatures(rest, result ++ [sig])
```

Comment:
@EdwinGuayacan @Odraxs you are building the Signatures here. The name of the function should reflect that